---
title: "Automated Analysis Pipeline for Viral Nanopore Sequencing Data"
author: "Ivanna Marchesini"
format: html
editor: visual
---

The `echo: false` option disables the printing of code (only output is displayed).

# Introduction

This pipeline performs automated processing and analysis of viral sequencing data generated using Oxford Nanopore Technologies (ONT). Steps include quality control, basecalling, read filtering, genome assembly, and result visualization.

# Pipeline Steps

## 1. Input Data

The first step in the pipeline involves downloading the `.pod5` raw signal files from the MinION device. This requires remote access to the sequencing machine via SSH.

### 1.1 Accessing the MinION via SSH

You can access the MinION device using the following credentials:

-   **Host name**: `MC-115199`
-   **IP address**: `10.212.1.44`

There are two user options for SSH login:

```         
ssh minit@10.212.1.44
```

password: minit

This command connects to the MinION device (run in terminal, not in Quarto)

**b. Personal User**

```         
 ssh imarches@10.212.1.44 
```

password: \[your personal password\]

after running the command above you will be asked to enter the password. Use a personal password or the default one.

This command connects to the MinION device (run in terminal, not in Quarto)

### 1.2 Navigate to the Sequencing Data in Minion Nanopore

```         
cd/data/run_20250803/no_sample_id/20250803_1227_MC-115199_AYL450_752e4334/pod5_pass/
```

List the available barcode folders and other directories:

```         
ls
```

| Output: barcode01 barcode02 barcode03 barcode07 barcode08 barcode09 barcode10 barcode11 barcode12 barcode13 barcode14 barcode15 barcode16 barcode17 barcode18 barcode19 barcode20 barcode21 barcode22 barcode23 barcode24 mixed unclassified

## Copy Data

Either copy the data with WinSCP to the computer or send it direct to Ramses using this following command:

```         
```

scp -r /data/run_20250803/no_sample_id/20250803_1227_MC-115199_AYL450_752e4334/pod5_pass imarches\@ramses4.itcc.uni-koeln.de:/projects/virology/nanopore/input/

## Log in to Ramses

Run the following command to log in:

```         
ssh imarches@ramses4.itcc.uni-koeln.de
```

You will be prompted to :

1.  Enter your SSH key passphrase
2.  Complete Duo two-factor authentication. You will see a prompt like this:

| Duo two-factor login for imarches
| Enter a passcode or select one of the following options:
| Duo Push to +XX XXX XXX9171 Passcode or option (1-1): (Hier you have to write the code that appears in your Duo app) Once authenticated successfully, you will see: Success. Logging you in...

Now are you logged into the server and can proceed with basecalling

# Ramses Set Up for basecalling

1.  Check job status

```         
sacct --format="JobID,ReqCPUS,UserCPU,Elapsed,ReqMem,MaxRSS,JobName%20,Stat"
```

# msa/freq files

1.  Resource Set-Up

    ```         
    mem=4gb 
    time=00:30:00 
    cores=1
    ```

These define SLURM job resources:

-   `--mem=4gb`

-   `--time=30 minutes`

-   `--cpus-per-task=1`

2.  Set imput path

    ```         
    pat0=nanopore/output/$project/barcode03/
    ```

3.  Loop

4.  Bild the command

```         
cmd='bamToFreq/bin/bamToFreq -q '$phredscore' '$pat0*.bam
```

Where:

-   `bamToFreq` is your frequency generator binary

-   `q $phredscore` sets the minimum base quality (like 20)git

-   The input BAM files are found under `$pat0`

Thatâ€™s the actual computation step to generate MSA/freq files. For each sample (barcode directory):

-   It runs `bamToFreq` on the contained BAM file(s)

-   Produces `.msa` and `.freq` files (as `bamToFreq` usually outputs both)

-   Logs everything into `/scratch/virology/logs/`

## Copiar los datos desde la computadora local a ramses (desde la terminal de la computadora local)

```         
scp -r "C:\Users\Viro_User\Documents\my_minion_sequencing\Input\run_20250803\no_sample_id\20250803_1227_MC-115199_AYL450_752e4334\pod5_pass" imarches@ramses4.itcc.uni-koeln.de:/projects/virology/nanopore/input/
```

Enter passphrase for key 'C:\Users\Viro\_User/.ssh/id_ed25519':

i had to change the name of the folder pod5_pass to run_20250803_ivi

```         
scp "C:\Users\Viro_User\Documents\my_minion_sequencing\Sample Sheed\some_sample_sheet_run_20250803_ivi.csv" imarches@ramses4.itcc.uni-koeln.de:/projects/virology/nanopore/input/run_20250803_ivi/
```

Enter passphrase for key 'C:\Users\Viro\_User/.ssh/id_ed25519':

to check if the sample sheet is okey or to change it manually:

```         
emacs some_sample_sheet_run_20250803_ivi.csv
```

flechas para moverse y escribir

ctrl X crtl S para guardar

ctrl X ctrl C para salir

copiar el sh a ramses para que la pipeline corra automaticamente

```         
scp "C:\Users\Viro_User\Documents\my_minion_sequencing\nanopore_basecalling_ivi.sh" imarches@ramses4.itcc.uni-koeln.de:/projects/virology/nanopore/input/
```

Enter passphrase for key 'C:\Users\Viro\_User/.ssh/id_ed25519':
